src/systems/hydrodynamics/Hydrodynamics.cc

```diff
@@ -78,14 +79,138 @@ class gz::sim::systems::HydrodynamicsPrivateData
   /// \brief Previous state.
   public: Eigen::VectorXd prevState;
 
+  /// \brief Previous derivative of the state
   public: Eigen::VectorXd prevStateDot;
 
-  /// Link entity
+  /// \brief Use current table if true
+  public: bool useCurrentTable {false};
+
+  /// \brief Current table xComponent
+  public: std::string axisComponents[3];
+
+  public: std::shared_ptr<gz::sim::components::EnvironmentalData>
+    gridField;
+
+  public: std::optional<gz::math::InMemorySession<double, double>> session[3];
+
+  /// \brief Link entity
   public: Entity linkEntity;
 
   /// \brief Ocean current callback
   public: void UpdateCurrent(const msgs::Vector3d &_msg);
 
+  /////////////////////////////////////////////////
+  /// \brief Set the current table
+  /// \param[in] _ecm - The Entity Component Manager
+  /// \param[in] _currTime - The current time
+  public: void SetCurrentTable(
+    const EntityComponentManager &_ecm,
+    const std::chrono::steady_clock::duration &_currTime)
+  {
+    _ecm.EachNew<components::Environment>([&](const Entity &/*_entity*/,
+      const components::Environment *_environment) -> bool
+    {
+      auto data = _environment->Data();
+      this->gridField = data;
+
+      for (std::size_t i = 0; i < 3; i++)
+      {
+        if (this->axisComponents[i] != "")
+        {
+
+          if (!data->frame.Has(this->axisComponents[i]))
+          {
+            gzwarn << "Environmental sensor could not find field "
+              << this->axisComponents[i] << "\n";
+            continue;
+          }
+
+          this->session[i] =
+            this->gridField->frame[this->axisComponents[i]].CreateSession();
+          if (!this->gridField->staticTime)
+          {
+            this->session[i] =
+              this->gridField->frame[this->axisComponents[i]].StepTo(
+                *this->session[i],
+                std::chrono::duration<double>(_currTime).count());
+          }
+
+          if(!this->session[i].has_value())
+          {
+            gzerr << "Exceeded time stamp." << std::endl;
+          }
+        }
+      }
+      return true;
+    });
+  }
+
+  /////////////////////////////////////////////////
+  /// \brief Retrieve the currernt from the data table
+  /// \param[in] _ecm - The Entity Component Manager
+  /// \param[in] _currTime - The current time
+  /// \param[in] _position - Position of the vehicle in world coordinates.
+  /// \return The current vector to be applied at this position and time.
+  public: math::Vector3d GetCurrentFromEnvironment(
+    const EntityComponentManager &_ecm,
+    const std::chrono::steady_clock::duration &_currTime,
+    math::Vector3d _position)
+  {
+    math::Vector3d current(0, 0, 0);
+
+    if (!this->gridField ||
+      !(this->session[0].has_value() ||
+        this->session[1].has_value() || this->session[1].has_value()))
```

Comment:
@arjo129 I think it should be
```suggestion
        this->session[1].has_value() || this->session[2].has_value()))
```